test/unit/test_Matrix.cc

```diff
@@ -0,0 +1,566 @@
+
+    LAVecRef u = ms.discretize(v);
+    ASSERT_EQ(vs[u][1], 0);
+    ASSERT_EQ(vs[u][2], 0);
+    ASSERT_EQ(vs[u][3], 0);
+    ASSERT_EQ(vs[u][4], 0);
+    ASSERT_EQ(vs[u][5], -1);
+    ASSERT_EQ(vs[u][6], 1);
+    ASSERT_EQ(vs[u][7], 1);
+    ASSERT_EQ(vs[u][8], -1);
+    ASSERT_EQ(vs[u][9], 1);
+    ASSERT_EQ(vs[u][10], -1);
+    ASSERT_EQ(vs[u][11], 0);
+    ASSERT_EQ(vs[u][12], 0);
+}
+
+TEST(Matrix_test, non_square)
+{
+    LAVecAllocator va;
+    ArithLogic logic{ArithLogic::ArithType::LIA};
+    LAVecStore vecStore(va, logic);
+    LAMatrixStore ms(vecStore);
+    MId U = ms.getNewMatrix(2,3); // Should be 2 columns, 3 rows
+    ms.MM(U, 1, 1) = 11;
+    ms.MM(U, 2, 1) = 21;
+    ms.MM(U, 1, 2) = 12;
+    ms.MM(U, 2, 2) = 22;
+    ms.MM(U, 1, 3) = 13;
+    ms.MM(U, 2, 3) = 23;
+    char* m_str = ms.print(U);
+    printf("The matrix:\n%s\n", m_str);
+    free(m_str);
+}
+
+TEST(gcd_test, gcd)
+{
+    FastRational p1(7177);
+    FastRational p2(7817);
+    FastRational p3(173);
+
+    FastRational q1 = p1*p3;
+    FastRational q2 = p2*p3;
+
+    ASSERT_EQ(gcd(p1, p2), 1);
+    ASSERT_EQ(gcd(p1, p3), 1);
+    ASSERT_EQ(gcd(p2, p3), 1);
+    ASSERT_EQ(gcd(q1, q2), p3);
+    ASSERT_EQ(gcd(q2, q1), p3);
+
+    FastRational p4(7919);
+    FastRational p5(7907);
+    FastRational q3 = p1*p2*p4;
+    FastRational q4 = p1*p2*p5;
+
+    ASSERT_EQ(gcd(q3, q4), p1*p2);
+
+    FastRational r1 = q3*q4;
+    FastRational r2 = q3*q4*p3;
+    ASSERT_EQ(gcd(r1, r2), q3*q4);
+}
+
+TEST(lcm_test, lcm)
+{
+    FastRational p1(2);
+    FastRational p2(6);
+
+    ASSERT_EQ(lcm(p1, p2), 6);
+    ASSERT_EQ(lcm(p2, p1), 6);
+
+    FastRational p3(7177);
+    FastRational p4(7817);
+    FastRational p5(7919);
+    FastRational p6(7907);
+
+    ASSERT_EQ(lcm(p3, p4), p3*p4);
+
+
+    FastRational q1 = p3*p4;
+    FastRational q2 = p5*p6;
+
+    ASSERT_EQ(lcm(q1, q2), p3*p4*p5*p6);
+    FastRational p7(4);
+    FastRational p8("2147483648");
+    FastRational q3 = p7*p8;
+    ASSERT_EQ(lcm(p7, p8), p8);
+    ASSERT_EQ(lcm(p8, p7), p8);
+
+
+}
+
+TEST(lcm_test, lcm_list)
+{
+    FastRational p1(2, 5);
+    FastRational p2(3, 7);
+    FastRational p3(5, 11);
+
+    std::vector<FastRational> l = {p1, p2, p3};
+    FastRational r = get_multiplicand(l);
+    ASSERT_EQ(r, 385);
+}
+
+TEST(Matrix_test, least_in_row) {
+    LAVecAllocator va;
+    ArithLogic logic{ArithLogic::ArithType::LIA};
+    LAVecStore vecStore(va, logic);
+    LAMatrixStore ms(vecStore);
+    MId U = ms.getNewMatrix(4, 4);
+    int u[4][4] = {{2, 3, 6, 2},
+                   {5, -6, -1, 6},
+                   {8, 3, -1, 1},
+                   {0, 0, 0, 0}};
+
+    for (int i = 0; i < 4; i++)
+        for (int j = 0; j < 4; j++)
+            ms.MM(U, i+1, j+1) = u[i][j];
+
+    ASSERT_EQ(ms.least_in_row(U, 1, 1), 4);
+    ASSERT_EQ(ms.least_in_row(U, 2, 2), 3);
+    ASSERT_EQ(ms.least_in_row(U, 3, 3), 4);
+    ASSERT_EQ(ms.least_in_row(U, 4, 1), 0);
+}
```

Comment:
And the same for `test/unit/test_Matrix.cc`

Reply:
Fixed.